File: javascript/49_Group_Anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    const res = new Map();

    for (let s of strs) {
        const count = new Array(26).fill(0);

        for (let c of s) {
            count[c.charCodeAt(0) - 'a'.charCodeAt(0)] ++;
        }
        res[count] ? res[count].push(s): res[count] =[s];  
    }
    return Object.values(res);
};

File: python/array/49_Group_Anagrams.py
'''
Time: O(26*m*n) = O(m*n)  -- m: len(strs); n: avg(len(each word))
'''
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # {}: throws a KeyError if you try to get an item with a key that is not currently in the dict. 
        # defaultdict: if you try to append to a nonexistent key, it will first create the key with the default value (an empty list)
        res = defaultdict(list)  # hashmap

        for s in strs:
            count = [0] * 26  # 26 letters in total

            for c in s:
                count[ord(c) - ord('a')] += 1  # idx: [0-25] 

            res[tuple(count)].append(s)  # in python, list can't be index, so use tuple

        return list(res.values())
